src/Raven.Server/Documents/AI/Embeddings/QueryEmbeddingsBatchingService.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Operations.AI;
+using Raven.Client.Util;
+using Raven.Server.Documents.ETL.Providers.AI;
+using Sparrow.Server.Logging;
+
+namespace Raven.Server.Documents.AI.Embeddings
+{
+    public sealed class QueryEmbeddingsBatchingService(AiIntegrationsController aiIntegrations) : IDisposable
+    {
+        private readonly SemaphoreSlim _globalConcurrencyLimiter = new(aiIntegrations.Database.Configuration.Ai.QueryEmbeddingsMaxConcurrentBatches);
+        private readonly RavenLogger _logger = aiIntegrations.Database.Loggers.GetLogger<QueryEmbeddingsBatchingService>();
+
+        private readonly ConcurrentDictionary<AiConnectionStringIdentifier, QueryEmbeddingsBatchingWorker> _batchWorkers = new();
+        // Flag for preventing the addition of new workers during disposal
+        private volatile bool _isDisposing;
+
+        public const string ShutdownMessage = $"{nameof(QueryEmbeddingsBatchingService)} is shutting down";
+
+        public ValueTask<ReadOnlyMemory<float>[]> GetEmbeddingAsync(AiConnectionStringIdentifier connectionStringId, IList<string> values, CancellationToken cancellationToken = default)
+        {
+            if (_isDisposing)
+            {
+                var tcs = new TaskCompletionSource<ReadOnlyMemory<float>[]>(TaskCreationOptions.RunContinuationsAsynchronously);
+                tcs.SetException(new ObjectDisposedException(ShutdownMessage));
+                return new ValueTask<ReadOnlyMemory<float>[]>(tcs.Task);
+            }
+
+            var batchWorker = _batchWorkers.GetOrAdd(connectionStringId, _ => CreateBatchWorker(connectionStringId));
```

Comment:
Doing it like this means that we don't need to allocate a lambda on each call.

```suggestion
            var batchWorker = _batchWorkers.GetOrAdd(connectionStringId, static id => CreateBatchWorker(id));
```

Reply:
done